torch/csrc/distributed/c10d/comm.h

```diff
@@ -13,4 +14,32 @@ void broadcast_coalesced(
     at::TensorList tensors,
     size_t buffer_size);
 
+class GradBucket {
+ public:
+  explicit GradBucket(std::vector<at::Tensor>& tensors);
+  std::vector<at::Tensor> get_tensors();
+
+ private:
+  std::vector<at::Tensor> tensors_;
+};
+
+struct CommHookInterface {
+ public:
+  virtual c10::intrusive_ptr<torch::jit::Future> operate(
+      const GradBucket& bucket) = 0;
+  virtual std::vector<at::Tensor> process_future(c10::IValue future_value) = 0;
+};
+
+class TORCH_API PythonCommHook : public CommHookInterface {
+ public:
+  PythonCommHook(py::object state, py::object hook);
+
+  c10::intrusive_ptr<torch::jit::Future> operate(
+      const GradBucket& bucket) override;
+  std::vector<at::Tensor> process_future(c10::IValue future_value) override;
+
+ private:
+  py::object state_;
+  py::object hook_;
```

Comment:
If this object takes the ownership of `py::object`, you will need to acquire GIL in its destructor and deref the `py::object`, otherwise, it would mess up PY internal states. See #38364 as an example.

Reply:
This might cause some trouble in the future, thank you for noticing it.